File: AIKeyWipe-Tauri/src-tauri/src/wipe_service.rs
use std::fs;
use std::io::Read;
use std::path::Path;
use regex::Regex;
use crate::models::{WipeTarget, WipeResult, ScannedFile};

/// 内置默认匹配模式
const DEFAULT_PATTERNS: &[&str] = &[
    "KEY=", "API_KEY=", "TOKEN=", "SECRET=", "PASSWORD=",
    "api_key:", "token:", "secret:", "_API_KEY",
];

/// 要扫描的配置文件扩展名
const CONFIG_EXTS: &[&str] = &[".env", ".envrc", ".yaml", ".yml", ".json", ".toml", ".conf", ".cfg", ".txt", ".ini", ".properties"];

/// 平台无关的通用跳过目录
const SKIP_DIRS_COMMON: &[&str] = &[
    ".git", "node_modules", ".cache", "Caches", "__pycache__",
    "venv", ".venv", ".Trash", ".rvm", ".nvm",
    ".oh-my-zsh", ".gem", ".cocoapods", ".m2",
    "Pods", "build", "dist", ".next", ".turbo",
];

/// macOS 特定跳过目录
/// 注意：Desktop/Downloads/Documents 不跳过，用户可能在这些目录下放项目文件
#[cfg(target_os = "macos")]
const SKIP_DIRS_PLATFORM: &[&str] = &[
    "Library", "Applications",
    "Public", "Music", "Pictures", "Movies",
];

/// Windows 特定跳过目录
/// 注意：Desktop/Documents/Downloads 不跳过，用户可能在这些目录下放项目文件
#[cfg(target_os = "windows")]
const SKIP_DIRS_PLATFORM: &[&str] = &[
    "AppData", "Application Data", "Local Settings",
    "Music", "Pictures", "Videos",
    "Public", "OneDrive", "3D Objects",
    "Contacts", "Favorites", "Links", "Saved Games", "Searches",
    "Windows", "System32", "Program Files", "Program Files (x86)",
    "ProgramData", "Recovery", "System Volume Information",
];

/// Linux 特定跳过目录
#[cfg(target_os = "linux")]
const SKIP_DIRS_PLATFORM: &[&str] = &[
    "Desktop", "Downloads", "Documents", "Music", "Pictures", "Videos",
    "Public", "snap", "flatpak", ".local", ".config",
];

/// 合并所有跳过目录
fn is_skip_dir(name: &str) -> bool {
    SKIP_DIRS_COMMON.contains(&name) || SKIP_DIRS_PLATFORM.contains(&name)
}

/// 合并内置 + 自定义模式
fn build_patterns(custom: &[String]) -> Vec<String> {
    let mut all: Vec<String> = DEFAULT_PATTERNS.iter().map(|s| s.to_string()).collect();
    for p in custom {
        let upper = p.to_uppercase();
        if !all.iter().any(|x| x.to_uppercase() == upper) {
            all.push(p.clone());
        }
    }
    all
}

/// 判断是否为 YAML 格式行（= 出现在第一个 : 之前说明是 .env 格式）
fn is_yaml_key(line: &str) -> bool {
    if let Some(pos) = line.find(':') {
        let before = &line[..pos];
        !before.contains('=')
    } else {
        false
    }
}

/// 清除文件中的 API Key
pub fn wipe_file(path: &str, custom_patterns: &[String]) -> Result<WipeResult, String> {
    let content = fs::read_to_string(path)
        .map_err(|e| format!("读取失败: {}", e))?;

    let all_lines: Vec<&str> = content.lines().collect();
    let patterns = build_patterns(custom_patterns);
    let mut modified_lines: Vec<String> = Vec::new();
    let mut wiped_count = 0;
    let mut skip_indented = false;

    for (i, line) in all_lines.iter().enumerate() {
        if skip_indented {
            let trimmed = line.trim();
            if !trimmed.is_empty() && line.starts_with(' ') {
                skip_indented = false;
                continue;
            }
            skip_indented = false;
        }

        let trimmed = line.trim();
        let leading_spaces: &str = &line[..line.len() - line.trim_start().len()];

        if trimmed.starts_with('#') || trimmed.starts_with("//") {
            modified_lines.push(line.to_string());
            continue;
        }

        let is_key_line = patterns.iter().any(|p| trimmed.contains(p.as_str()));
        if !is_key_line {
            modified_lines.push(line.to_string());
            continue;
        }

        if !custom_patterns.is_empty() {
            let trimmed2 = trimmed;
            if custom_patterns.iter().any(|cp| trimmed2 == cp.as_str() || trimmed2 == cp.trim()) {
                modified_lines.push(String::new());
                wiped_count += 1;
                continue;
            }
        }

        let yaml = is_yaml_key(trimmed);
        let has_inline_value = if yaml {
            let re = Regex::new(r"^[^:]*:\s*").unwrap();
            let after = re.replace(trimmed, "");
            !after.is_empty() && after != "\"\""
        } else {
            let re = Regex::new(r"^[^=]*=").unwrap();
            let after = re.replace(trimmed, "");
            !after.is_empty()
        };

        if trimmed.contains("export ") {
            let re = Regex::new(r"^(export\s+[A-Za-z_][A-Za-z0-9_]*\s*=).*").unwrap();
            let clean = re.replace(trimmed, "$1");
            modified_lines.push(clean.to_string());
        } else if yaml {
            if has_inline_value {
                let re = Regex::new(r"^([a-zA-Z_][a-zA-Z0-9_]*\s*:\s*).*").unwrap();
                let clean = re.replace(trimmed, "$1\"\"");
                modified_lines.push(format!("{}{}", leading_spaces, clean));
            } else {
                let re = Regex::new(r":.*").unwrap();
                let clean = re.replace(trimmed, ": \"\"");
                modified_lines.push(format!("{}{}", leading_spaces, clean));
                if i + 1 < all_lines.len() {
                    let next = all_lines[i + 1];
                    if next.starts_with(' ') && !next.trim().is_empty() {
                        skip_indented = true;
                    }
                }
            }
        } else {
            if has_inline_value {
                let re = Regex::new(r"^([A-Za-z_][A-Za-z0-9_]*\s*=).*").unwrap();
                let clean = re.replace(trimmed, "$1");
                modified_lines.push(clean.to_string());
            } else {
                modified_lines.push(line.to_string());
            }
        }
        wiped_count += 1;
    }

    let new_content = modified_lines.join("\n");
    fs::write(path, &new_content)
        .map_err(|e| format!("写入失败: {}", e))?;

    let file_name = std::path::Path::new(path)
        .file_name().unwrap_or_default().to_string_lossy().to_string();

    Ok(WipeResult {
        target_id: None,
        target_name: file_name,
        status: "success".to_string(),
        message: format!("已清除 {} 个 Key", wiped_count),
    })
}

/// 批量执行清除
pub fn execute_wipe(targets: &[WipeTarget]) -> Vec<WipeResult> {
    let mut results = Vec::new();
    for target in targets {
        if !target.enabled {
            results.push(WipeResult {
                target_id: Some(target.id.clone()),
                target_name: target.name.clone(),
                status: "skipped".to_string(),
                message: "已禁用，跳过".to_string(),
            });
            continue;
        }
        let path = target.expanded_path();
        match wipe_file(&path, &target.custom_patterns) {
            Ok(mut r) => {
                r.target_id = Some(target.id.clone());
                r.target_name = target.name.clone();
                results.push(r);
            }
            Err(e) => {
                results.push(WipeResult {
                    target_id: Some(target.id.clone()),
                    target_name: target.name.clone(),
                    status: "failed".to_string(),
                    message: e,
                });
            }
        }
    }
    results
}

/// 扫描 ~/ 下的配置文件，返回含 API Key 的文件列表
pub fn scan_for_keys(progress: &dyn Fn(&str)) -> Vec<ScannedFile> {
    let mut results = Vec::new();
    let home = match dirs::home_dir() {
        Some(h) => h,
        None => {
            progress("错误: 无法获取用户家目录路径");
            return results;
        }
    };

    let home_str = home.to_string_lossy().to_string();
    progress(&format!("开始扫描: {}", home_str));

    let patterns = DEFAULT_PATTERNS;

    #[cfg(target_os = "windows")]
    let max_depth = 7;
    #[cfg(not(target_os = "windows"))]
    let max_depth = 6;

    scan_dir_recursive(&home, 0, max_depth, patterns, &mut results, progress);

    #[cfg(target_os = "windows")]
    {
        let mut paths_to_check = vec![
            home.join(".env"),
            home.join(".envrc"),
            home.join(".gitconfig"),
        ];
        if let Some(config) = dirs::config_dir() {
            paths_to_check.push(config.join("pip").join("pip.conf"));
            paths_to_check.push(config.join("npmrc"));
            paths_to_check.push(config.join(".buckconfig"));
        }
        for p in &paths_to_check {
            if p.is_file() { check_file(p, patterns, &mut results, progress); }
        }
        if let Some(config) = dirs::config_dir() {
            if let Ok(entries) = fs::read_dir(&config) {
                for entry in entries.flatten() {
                    let path = entry.path();
                    if path.is_file() { check_file(&path, patterns, &mut results, progress); }
                }
            }
        }
    }

    progress(&format!("扫描完成，共发现 {} 个文件", results.len()));
    results
}

/// 递归扫描目录（最大深度限制）
fn scan_dir_recursive(
    dir: &Path,
    depth: usize,
    max_depth: usize,
    patterns: &[&str],
    results: &mut Vec<ScannedFile>,
    progress: &dyn Fn(&str),
) {
    if depth > max_depth {
        return;
    }

    if depth >= 2 {
        progress(&format!("扫描: {}", dir.display()));
    }

    let entries = match fs::read_dir(dir) {
        Ok(e) => e,
        Err(e) => {
            progress(&format!("⚠ 无法读取目录 ({}): {}", dir.display(), e));
            return;
        }
    };

    for entry in entries.flatten() {
        let path = entry.path();

        if path.is_symlink() {
            continue;
        }

        if path.is_dir() {
            let name = path.file_name()
                .unwrap_or_default()
                .to_string_lossy()
                .to_string();

            // 只靠 SKIP_DIRS 过滤，不跳过任何隐藏目录
            if is_skip_dir(&name) {
                continue;
            }

            scan_dir_recursive(&path, depth + 1, max_depth, patterns, results, progress);
            continue;
        }

        if path.is_file() {
            check_file(&path, patterns, results, progress);
        }
    }
}

fn check_file(
    path: &Path,
    patterns: &[&str],
    results: &mut Vec<ScannedFile>,
    progress: &dyn Fn(&str),
) {
    // 取消扩展名限制，扫描所有文件
    // 跳过明显不可读的目录（缓存/临时）
    let path_buf = path.to_path_buf();
    let path_str = path_buf.to_string_lossy();
    if path_str.contains("/.Trash/") || path_str.contains("/.cache/") {
        return;
    }

    progress(&format!("扫描: {}", path_buf.file_name().unwrap_or_default().to_string_lossy()));

    let mut file = match fs::File::open(path) {
        Ok(f) => f,
        _ => return,
    };
    let mut buf = vec![0u8; 10240];
    let n = file.read(&mut buf).unwrap_or(0);
    if n == 0 { return; }
    let content = String::from_utf8_lossy(&buf[..n]);

    let has_key = content.lines().any(|line| {
        let t = line.trim();
        if t.starts_with('#') || t.is_empty() { return false; }
        patterns.iter().any(|p| t.contains(p))
    });

    if has_key {
        let name = format!("{}/{}",
            path_buf.parent().and_then(|p| p.file_name()).map(|n| n.to_string_lossy()).unwrap_or_default(),
            path_buf.file_name().map(|n| n.to_string_lossy()).unwrap_or_default());
        let full_path = path_buf.to_string_lossy().to_string();
        if !results.iter().any(|r| r.path == full_path) {
            results.push(ScannedFile { path: full_path, name });
        }
    }
}

File: AIKeyWipe/Sources/AIKeyWipe/WipeService.swift
import Foundation

/// 执行清除操作的服务
class WipeService {

    private let fileManager = FileManager.default
    private let dateFormatter: DateFormatter = {
        let f = DateFormatter()
        f.dateFormat = "yyyyMMdd_HHmmss"
        return f
    }()

    var backupBaseURL: URL {
        let home = fileManager.homeDirectoryForCurrentUser
        let backups = home.appendingPathComponent("AI-Key-Wipe/backups")
        try? fileManager.createDirectory(at: backups, withIntermediateDirectories: true)
        return backups
    }

    // MARK: - 内置默认匹配模式
    private static let defaultPatterns = [
        "KEY=", "API_KEY=", "TOKEN=", "SECRET=", "PASSWORD=",
        "api_key:", "token:", "secret:", "_API_KEY"
    ]

    /// 合并内置 + 自定义模式（自定义优先）
    private func buildPatterns(custom: [String]) -> [String] {
        var all = Self.defaultPatterns
        for p in custom {
            let upper = p.uppercased()
            if !all.contains(where: { $0.uppercased() == upper }) {
                all.append(p)
            }
        }
        return all
    }

    // MARK: - 扫描文件中的 API Key
    func scanFile(at path: String, customPatterns: [String] = []) -> [String] {
        let url = URL(fileURLWithPath: path)
        guard fileManager.fileExists(atPath: url.path) else { return [] }

        guard let content = try? String(contentsOf: url, encoding: .utf8) else { return [] }

        var found: [String] = []

        let lines = content.components(separatedBy: .newlines)
        for line in lines {
            let trimmed = line.trimmingCharacters(in: .whitespaces)
            // 跳过注释
            if trimmed.hasPrefix("#") || trimmed.hasPrefix("//") { continue }

            // 匹配 KEY= / TOKEN= / SECRET= / PASSWORD= 模式
            let patterns = buildPatterns(custom: customPatterns)
            for pattern in patterns {
                if trimmed.contains(pattern) {
                    let masked = trimmed.replacingOccurrences(
                        of: "=.*$", with: "=******",
                        options: .regularExpression
                    )
                    found.append(masked)
                    break
                }
            }
        }
        return found
    }

    // MARK: - 清除文件中的 Key
    func wipeFile(at path: String, customPatterns: [String] = [], targetId: UUID? = nil, backup: Bool = true) -> WipeResult? {
        let url = URL(fileURLWithPath: path)
        let fileName = url.lastPathComponent
        let dirName = url.deletingLastPathComponent().lastPathComponent
        let displayName = "\(dirName)/\(fileName)"

        guard fileManager.fileExists(atPath: url.path) else {
            return nil
        }

        // 1. 备份
        if backup {
            let dateStr = dateFormatter.string(from: Date())
            let backupDir = backupBaseURL.appendingPathComponent(dateStr)
            let dest = backupDir.appendingPathComponent("\(dirName)-\(fileName)")

            do {
                try fileManager.createDirectory(at: backupDir, withIntermediateDirectories: true)
                try fileManager.copyItem(at: url, to: dest)
            } catch {
                return WipeResult(
                    targetId: targetId,
                    targetName: displayName,
                    status: .failed,
                    message: "备份失败: \(error.localizedDescription)",
                    timestamp: Date()
                )
            }
        }

        do {
            var content = try String(contentsOf: url, encoding: .utf8)
            let allLines = content.components(separatedBy: .newlines)
            var modifiedLines: [String] = []
            var wipedCount = 0
            var skipIndentedValue = false  // 标记：下一行的缩进值是上一行 YAML key 的值，需要跳过

            for (i, line) in allLines.enumerated() {
                // ── 如果上一条是 YAML key: 且值在下一行，跳过本行 ──
                if skipIndentedValue {
                    let trimmed = line.trimmingCharacters(in: .whitespaces)
                    if !trimmed.isEmpty && line.hasPrefix(" ") {
                        // 跳过这行（它是 key: 的换行值）
                        skipIndentedValue = false
                        continue
                    }
                    skipIndentedValue = false
                    // 不跳过——它不是缩进值，继续正常处理
                }

                let trimmed = line.trimmingCharacters(in: .whitespaces)
                let leadingSpaces = line.prefix(while: { $0 == " " })

                // 跳过已注释的行
                if trimmed.hasPrefix("#") || trimmed.hasPrefix("//") {
                    modifiedLines.append(line)
                    continue
                }

                // ── 检查是否包含 Key/Token/Secret ──
                let keyPatterns = buildPatterns(custom: customPatterns)
                var isKeyLine = false
                for pattern in keyPatterns {
                    if trimmed.contains(pattern) {
                        isKeyLine = true
                        break
                    }
                }

                if !isKeyLine {
                    modifiedLines.append(line)
                    continue
                }

                // ── 自定义模式精确匹配整行 → 整行删除 ──
                // 如果用户设置的某个自定义模式与整行内容完全相同，直接删除该行
                if !customPatterns.isEmpty {
                    let isExactMatch = customPatterns.contains { cp in
                        trimmed == cp || trimmed == cp.trimmingCharacters(in: .whitespaces)
                    }
                    if isExactMatch {
                        modifiedLines.append("")  // 空行 = 删除该行
                        wipedCount += 1
                        continue
                    }
                }

                // ── 判断是否多行值（key: 后面没值，值在下一行） ──
                // 注意：只有 YAML 用 ":" 做 key-value 分隔符，.env 用 "="
                // 如果 "=" 出现在第一个 ":" 之前，说明这是 .env 格式而非 YAML
                let isYamlKey: Bool = {
                    if let colonRange = trimmed.range(of: ":") {
                        let beforeColon = trimmed[..<colonRange.lowerBound]
                        return !beforeColon.contains("=")
                    }
                    return false
                }()
                let hasInlineValue: Bool = {
                    if isYamlKey {
                        // 冒号后有非空内容才算有行内值
                        let afterColon = trimmed.replacingOccurrences(
                            of: #"^[^:]*:\s*"#, with: "",
                            options: .regularExpression
                        )
                        return !afterColon.isEmpty && afterColon != "\"\""
                    }
                    // KEY= 格式，等号后有内容才算有行内值
                    let afterEq = trimmed.replacingOccurrences(
                        of: #"^[^=]*="#, with: "",
                        options: .regularExpression
                    )
                    return !afterEq.isEmpty
                }()

                // ── 生成清除后的行 ──
                if trimmed.contains("export ") {
                    // export KEY=value → export KEY=
                    let cleanLine = trimmed.replacingOccurrences(
                        of: #"^(export\s+[A-Za-z_][A-Za-z0-9_]*\s*=).*"#,
                        with: "$1",
                        options: .regularExpression
                    )
                    modifiedLines.append(cleanLine)
                } else if isYamlKey {
                    // YAML 格式 — 清空值
                    if hasInlineValue {
                        // api_key: value → api_key: ""（保留原始缩进）
                        let cleanLine = trimmed.replacingOccurrences(
                            of: #"^([a-zA-Z_][a-zA-Z0-9_]*\s*:\s*).*"#,
                            with: "$1\"\"",
                            options: .regularExpression
                        )
                        modifiedLines.append("\(String(leadingSpaces))\(cleanLine)")
                    } else {
                        // api_key:\n    value → api_key: ""（并标记跳过下一行）
                        let indent = String(leadingSpaces)
                        modifiedLines.append("\(indent)\(trimmed.replacingOccurrences(of: ":.*", with: ": \"\"", options: .regularExpression))")
                        // 检查下一行是否是缩进的值
                        if i + 1 < allLines.count {
                            let next = allLines[i + 1]
                            if next.hasPrefix(" ") && !next.trimmingCharacters(in: .whitespaces).isEmpty {
                                skipIndentedValue = true
                            }
                        }
                    }
                } else {
                    // KEY=value → KEY=
                    if hasInlineValue {
                        let cleanLine = trimmed.replacingOccurrences(
                            of: #"^([A-Za-z_][A-Za-z0-9_]*\s*=).*"#,
                            with: "$1",
                            options: .regularExpression
                        )
                        modifiedLines.append(cleanLine)
                    } else {
                        // 等号后面没值，但变量名本身匹配了 KEY 模式——保留原行
                        modifiedLines.append(line)
                    }
                }
                wipedCount += 1
            }

            let newContent = modifiedLines.joined(separator: "\n")
            try newContent.write(to: url, atomically: true, encoding: .utf8)

            return WipeResult(
                targetId: targetId,
                targetName: displayName,
                status: .success,
                message: "已清除 \(wipedCount) 个 Key",
                timestamp: Date()
            )

        } catch {
            return WipeResult(
                targetId: targetId,
                targetName: displayName,
                status: .failed,
                message: "写入失败: \(error.localizedDescription)",
                timestamp: Date()
            )
        }
    }

    // MARK: - 特殊处理：Shell 历史
    func wipeShellHistory(at path: String, targetId: UUID? = nil, backup: Bool = true) -> WipeResult? {
        let url = URL(fileURLWithPath: path)
        let fileName = url.lastPathComponent

        guard fileManager.fileExists(atPath: url.path) else { return nil }

        if backup {
            let dateStr = dateFormatter.string(from: Date())
            let backupDir = backupBaseURL.appendingPathComponent(dateStr)
            let dest = backupDir.appendingPathComponent(fileName)
            do {
                try fileManager.createDirectory(at: backupDir, withIntermediateDirectories: true)
                try fileManager.copyItem(at: url, to: dest)
            } catch {
                return WipeResult(
                    targetId: targetId,
                    targetName: fileName,
                    status: .failed,
                    message: "备份失败: \(error.localizedDescription)",
                    timestamp: Date()
                )
            }
        }

        do {
            let content = try String(contentsOf: url, encoding: .utf8)
            let lines = content.components(separatedBy: .newlines)

            // 过滤掉包含 API Key 的行
            let filteredLines = lines.filter { line in
                let upper = line.uppercased()
                return !upper.contains("API_KEY") &&
                       !upper.contains("APIKEY") &&
                       !upper.contains("OPENAI") &&
                       !upper.contains("ANTHROPIC") &&
                       !upper.contains("DEEPSEEK") &&
                       !upper.contains("GEMINI") &&
                       !upper.contains("SECRET")
            }

            let removed = lines.count - filteredLines.count
            let newContent = filteredLines.joined(separator: "\n")
            try newContent.write(to: url, atomically: true, encoding: .utf8)

            return WipeResult(
                targetId: targetId,
                targetName: fileName,
                status: .success,
                message: "已从 Shell 历史中移除 \(removed) 条敏感命令",
                timestamp: Date()
            )
        } catch {
            return WipeResult(
                targetId: targetId,
                targetName: fileName,
                status: .failed,
                message: "处理失败: \(error.localizedDescription)",
                timestamp: Date()
            )
        }
    }

    // MARK: - 批量执行
    func executeWipe(on targets: [WipeTarget], backup: Bool = true, progressHandler: (Double, String) -> Void) -> [WipeResult] {
        var results: [WipeResult] = []
        let total = Double(targets.count)
        let dateStr = dateFormatter.string(from: Date())
        let backupRoot = backupBaseURL.appendingPathComponent(dateStr)
        try? fileManager.createDirectory(at: backupRoot, withIntermediateDirectories: true)

        // 保存本次操作的配置清单到备份目录
        let manifestURL = backupRoot.appendingPathComponent("_manifest.txt")
        let manifest = targets.enumerated().map { i, t in
            "\(i+1). \(t.name) → \(t.filePath)"
        }.joined(separator: "\n")
        try? manifest.write(to: manifestURL, atomically: true, encoding: .utf8)

        for (index, target) in targets.enumerated() {
            let progress = Double(index) / total
            progressHandler(progress, "正在处理: \(target.name)…")

            guard target.enabled else {
                results.append(WipeResult(
                    targetId: target.id,
                    targetName: target.name,
                    status: .skipped,
                    message: "已禁用，跳过",
                    timestamp: Date()
                ))
                continue
            }

            let path = target.expandedPath

            // 根据文件类型选择不同的处理方式
            if path.hasSuffix("_history") || path.hasSuffix(".zsh_history") || path.hasSuffix(".bash_history") {
                if let result = wipeShellHistory(at: path, targetId: target.id, backup: backup) {
                    results.append(result)
                } else {
                    results.append(WipeResult(
                        targetId: target.id,
                        targetName: target.name,
                        status: .skipped,
                        message: "文件不存在: \(path)",
                        timestamp: Date()
                    ))
                }
            } else if path.hasSuffix("history.jsonl") || path.hasSuffix(".hermes_history") {
                // JSONL 历史文件 — 直接清空
                let url = URL(fileURLWithPath: path)
                if fileManager.fileExists(atPath: path) {
                    // 备份
                    if backup {
                        let dest = backupRoot.appendingPathComponent(url.lastPathComponent)
                        try? fileManager.copyItem(at: url, to: dest)
                    }
                    // 清空
                    try? "".write(to: url, atomically: true, encoding: .utf8)
                    results.append(WipeResult(
                        targetId: target.id,
                        targetName: target.name,
                        status: .success,
                        message: "已清空历史文件",
                        timestamp: Date()
                    ))
                } else {
                    results.append(WipeResult(
                        targetId: target.id,
                        targetName: target.name,
                        status: .skipped,
                        message: "文件不存在",
                        timestamp: Date()
                    ))
                }
            } else {
                if let result = wipeFile(at: path, customPatterns: target.customPatterns, targetId: target.id, backup: backup) {
                    results.append(result)
                } else {
                    results.append(WipeResult(
                        targetId: target.id,
                        targetName: target.name,
                        status: .skipped,
                        message: "文件不存在: \(path)",
                        timestamp: Date()
                    ))
                }
            }
        }

        progressHandler(1.0, "清除完成！")
        return results
    }

    // MARK: - 自动扫描文件系统中的 API Key 文件
    /// 扫描 ~/ 下的 .env 和配置文件，返回发现的文件路径和推荐名称
    func scanForApiKeyFiles(progressHandler: (String) -> Void) -> [(path: String, name: String)] {
        var results: [(String, String)] = []
        let home = fileManager.homeDirectoryForCurrentUser
        let patterns = Self.defaultPatterns

        // 跳过的目录
        let skipDirs: Set<String> = [
            "Library", "Applications", "Desktop", "Downloads", "Documents",
            "Public", "Music", "Pictures", "Movies", ".Trash", ".git",
            "node_modules", ".cache", "Caches", ".rvm", ".nvm",
            ".oh-my-zsh", ".gem", ".cocoapods", ".m2",
            "venv", ".venv", "__pycache__", "Pods"
        ]

        // 要检查的文件扩展名
        let configExts: Set<String> = [".env", ".envrc", ".yaml", ".yml", ".json", ".toml", ".conf", ".cfg"]

        guard let enumerator = fileManager.enumerator(
            at: home,
            includingPropertiesForKeys: [.isDirectoryKey, .isRegularFileKey],
            options: [.skipsPackageDescendants]
        ) else { return [] }

        var depthLimit = 4
        for case let url as URL in enumerator {
            guard let resourceValues = try? url.resourceValues(forKeys: [.isDirectoryKey, .isRegularFileKey]) else { continue }

            if resourceValues.isDirectory == true {
                // 跳过不需要的目录
                if skipDirs.contains(url.lastPathComponent) {
                    enumerator.skipDescendants()
                    continue
                }
                // 深度限制
                let relativePath = url.path.replacingOccurrences(of: home.path, with: "")
                let depth = relativePath.components(separatedBy: "/").count
                if depth > depthLimit {
                    enumerator.skipDescendants()
                    continue
                }
                continue
            }

            guard resourceValues.isRegularFile == true else { continue }
            let ext = "." + url.pathExtension.lowercased()
            guard configExts.contains(ext) || url.lastPathComponent == ".env" else { continue }

            progressHandler("扫描: \(url.lastPathComponent)")

            // 只检查前 100 行
            guard let fileHandle = try? FileHandle(forReadingFrom: url) else { continue }
            let chunk = fileHandle.readData(ofLength: 10_240)
            fileHandle.closeFile()
            guard !chunk.isEmpty, let content = String(data: chunk, encoding: .utf8) else { continue }

            let lines = content.components(separatedBy: CharacterSet.newlines)
            var hasKey = false
            for line in lines {
                let t = line.trimmingCharacters(in: CharacterSet.whitespaces)
                if t.hasPrefix("#") || t.isEmpty { continue }
                for p in patterns {
                    if t.contains(p) {
                        hasKey = true
                        break
                    }
                }
                if hasKey { break }
            }

            if hasKey {
                let dirName = url.deletingLastPathComponent().lastPathComponent
                let fullPath = url.path
                let name = "\(dirName)/\(url.lastPathComponent)"
                // 去重
                if !results.contains(where: { $0.0 == fullPath }) {
                    results.append((fullPath, name))
                }
            }
        }
        return results
    }
}
